src/cascadia/WindowsTerminal/IslandWindow.cpp

```diff
@@ -967,34 +967,143 @@ void IslandWindow::SetGlobalHotkeys(const std::vector<winrt::Microsoft::Terminal
+// - down: if true, increase the height from top to bottom. otherwise, decrease
+//   the height, from bottom to top.
+// Return Value:
+// - <none>
+void IslandWindow::_doSlideAnimation(const uint32_t dropdownDuration, const bool down)
+{
+    til::rectangle fullWindowSize{ GetWindowRect() };
+    const double fullHeight = fullWindowSize.height<double>();
+
+    const double animationDuration = dropdownDuration; // use floating-point math throughout
+    const auto start = std::chrono::system_clock::now();
+
+    // Do at most dropdownDuration frames. After that, just bail straight to the
+    // final state.
+    for (uint32_t i = 0; i < dropdownDuration; i++)
+    {
+        const auto end = std::chrono::system_clock::now();
+        const auto millis = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
+        const double dt = ::base::saturated_cast<double>(millis.count());
+
+        if (dt > animationDuration)
+        {
+            break;
+        }
+
+        // If going down, increase the height over time. If going up, decrease the height.
+        const double currentHeight = ::base::saturated_cast<double>(
+            down ? ((dt / animationDuration) * fullHeight) :
+                   ((1.0 - (dt / animationDuration)) * fullHeight));
+
+        wil::unique_hrgn rgn{ CreateRectRgn(0,
+                                            0,
+                                            fullWindowSize.width<int>(),
+                                            ::base::saturated_cast<int>(currentHeight)) };
+
+        SetWindowRgn(_interopWindowHandle, rgn.get(), true);
+
+        // Go immediately into another frame. This prevents the window from
+        // doing anything else (tearing our state). A Sleep() here will cause a
+        // weird stutter, and causes the animation to not be as smooth.
+    }
+
+    // Reset the window.
+    SetWindowRgn(_interopWindowHandle, nullptr, true);
+}
+
+void IslandWindow::_dropdownWindow(const uint32_t dropdownDuration)
+{
+    // First, restore the window. SetWindowPlacement has a fun undocumented
+    // piece of functionality where it will restore the window position
+    // _without_ the animation, so use that instead of ShowWindow(SW_RESTORE).
+    WINDOWPLACEMENT wpc;
+    wpc.length = sizeof(WINDOWPLACEMENT);
+    GetWindowPlacement(_window.get(), &wpc);
+    wpc.showCmd = SW_RESTORE;
+    SetWindowPlacement(_window.get(), &wpc);
+
+    // Now that we're visible, animate the dropdown.
+    _doSlideAnimation(dropdownDuration, true);
+}
+
+void IslandWindow::_slideUpWindow(const uint32_t dropdownDuration)
+{
+    // First, animate the window sliding up.
+    _doSlideAnimation(dropdownDuration, false);
+
+    // Then, use SetWindowPlacement to minimize without the animation.
+    WINDOWPLACEMENT wpc;
+    wpc.length = sizeof(WINDOWPLACEMENT);
+    GetWindowPlacement(_window.get(), &wpc);
+    wpc.showCmd = SW_MINIMIZE;
+    SetWindowPlacement(_window.get(), &wpc);
+}

```

Comment:
These two seem very similar. If you add a second param `const bool down` (like in `_doSlideAnimation`), you can cut down on this repeated code.